geode-core/src/main/java/org/apache/geode/distributed/internal/membership/gms/fd/GMSHealthMonitor.java

```diff
@@ -976,14 +975,17 @@ private void stopServices() {
     }
 
     if (serverSocketExecutor != null) {
-      if (serverSocket != null && !serverSocket.isClosed()) {
-        try {
-          serverSocket.close();
-          serverSocket = null;
-          logger.info("GMSHealthMonitor server socket is closed in stopServices().");
-        } catch (IOException e) {
-          logger.trace("Unexpected exception", e);
+      if (serverSocket != null) {
```

Comment:
If there are two threads going after this socket maybe a synchronization is needed.  If you don't want to do that I think you should revert the _volatile_ change and read the value of serverSocket into a local variable.  Then do the null check, etc., on the local variable.

Reply:
The socket actually handles it's synchronization fine.  The volatile was there because the TcpServer thread was also nulling this variable (though not actually using it aside from that).  There doesn't seem to be any reason why that thread should be doing that though, so I removed that line (old 734) and left the new line 988 as the sole place where this gets nulled...although apparently I didn't delete line 982 like I thought I did so I'll do that.  It seems like the volatile is no longer useful, though I suppose there could be a race if one thread was calling started() while another was calling stopServices(), though honestly I think we'd have other issues at that point.  Still, I'll add the volatile back just to make sure I'm not accidentally breaking something I'm not currently seeing.

Reply:
It looks like `serverSocket` is written in `started()` and read/written in `stopServices()`.  

Marking `serverSocket` as _volatile_ ensures visibility of the write to other threads.  It's totally possible that w/o volatile the write is never seen in the absence of external synchronization (outside of this class).

It doesn't provide mutual exclusion for concurrent executions of `stopServices()`.  Is that a possibility?

My preference is to encapsulate synchronization of mutable state (e.g. synchronizing all reads/write of mutable state) to avoid complexity and data races.

Reply:
Yeah, leaving 'serverSocket' as volatile is certainly less risky without having a better understanding of this code.

'stopServices()' does not appear to be reentrant, so I'm not concerned about concurrent executions of it.  Honestly, nulling out 'serverSocket' does not even seem to be a useful step here, aside from not wanting to change this code more than necessary to fix the race.  Should I just remove this so that the value is only modified in the 'started()' call?